### PR TITLE
feat(macro): record a macro on the keyboard, and give every slot a stock look

### PR DESCRIPTION
**Pull request by thpoll83**
## Summary

Three things, in one branch because they are one feature from the user's side.

**1. Macros have a home on the board.** `QK_MACRO_0..11` sit on the utility layer where `F13..F24` used to be, with Shift reaching M12..M15. `poly_macro_banked_id()` is one implementation shared by the action path and the render path, so a keycap can never show M13 while the key plays M1 — the action reads the live `get_mods()`, the legend the synced snapshot, the same deliberate asymmetry `KC_GLYPH_SIZE_UP` uses.

⚠️ **`F13..F24` lose their default home.** A user who wants them back assigns them from the layout editor; this needs saying in the release notes.

**2. An unclaimed slot ships a stock look** — the Mayan numeral for its own index over the caption `Macro`. Without it a keyboard that has never met the host app shows sixteen keycaps distinguished only by "M0".."M15", which is the twelve-keys-that-look-alike problem the displays exist to solve.

Mayan is the only numeral system that fits: base-20, so 0..15 are each a *single* glyph; it has a real glyph for **zero** (the shell, U+1D2E0) rather than an absence, which is what lets the set reach M0 at all; and bar-and-dot is what a 1-bit 72×40 panel draws well. Nothing already in the pack covered them — measured, 0 of 20 codepoints resolved — so this adds `NotoSansMayanNumerals` (OFL, 50 KB) and reships the `symbol` bundle (v8 → v9, 37,200 → 38,976 B in a 96 KB slot). The entry sits at the very **end** of `fonts.yaml`'s `fonts` list, not the end of the symbols block: the category picks the bundle, the list position picks the global index, and appending it after the symbols entries would have shifted the whole `fantasy` bundle and forced a second reship for a font nothing else touched.

**3. Recording, with no host app.** Tap `KC_MACRO_REC` on the utility layer → the board becomes a slot picker → tap a macro key → type → tap REC again.

- Capture is a `host_driver_t` shim over the HID reports, so a macro plays back what the keyboard actually **sent** — after every layer, mod-tap and combo resolved — rather than what the matrix did. ⚠️ The shim is installed from **housekeeping**, not `keyboard_post_init_user()`: `protocol_post_init()` runs after the post_init hooks and installs the USB driver, so a shim set there is overwritten and the recording silently captures nothing.
- The diff is pure in `base/macro_record.c` (no quantum.h, no EEPROM, no timer) — the same seam as `base/fw_up_verdict.c` and for the same reason. `keys[6]` is an unordered **set**, so it asks whether each code is present in the other report rather than comparing positions; and within one diff releases are emitted before presses, or playback can hold a modifier the user had already lifted and type the shifted character.
- Steps land in a 192 B RAM buffer and the splice is pumped a chunk per housekeeping pass. A write per keystroke is a wear-levelling journal append, and the consolidation erase it triggers is the documented mechanism behind the "slave becomes unresponsive" bug — mid-recording is exactly when it would land.
- `rec_state`/`rec_slot` are synced because the slave draws its own half of the picker. The byte count deliberately is not: it moves per keystroke, so syncing it would buy a bridge frame per keypress. The master's panel shows `48/192 B` and the slave shows the stop hint in its place.
- The status OLED is the only indicator the gesture has (split42 has no RGB matrix, and the keycaps are busy showing what is being typed), so its branch sits above `DISP_IDLE` and housekeeping holds `update_performed()` while recording.

**Two bugs found and fixed in review rounds on hardware**, both worth knowing:

- `KC_MACRO_REC` was **rendering a blank keycap** — the keycode was in both keymaps, the action path and the OLED, but neither legend switch had a case for it. A missing legend is a missing `case`, not a compile error, so every build and the whole lint job stayed green while the key was invisible.
- A stock-look change is **invisible on an already-flashed board**: the look is stamped into EEPROM the first time a slot is found empty, so `slot_unclaimed()` is false from then on. `slot_holds_legacy_seed()` re-seeds a slot whose body is **empty** *and* whose record matches, byte for byte, what the previous scheme wrote — so a caption someone typed themselves is never touched, and neither is a slot holding a real macro. It is marked deletable once no field board predates the change.

Design doc and the alternatives weighed against this gesture: `keyboards/polykybd/MACRO_RECORD_DESIGN.md`.

## Version bump label

`bump:minor` — new feature, backwards compatible. No `PROTOCOL_VERSION` change: the macro commands (36/37/38) are already v15, and the `['G'][u16 state_generation]` block this branch adds to the `GET_ID` reply is tag-led and optional, so its absence means "no generation available" and an older host stops at the `V` block as before.

## Testing

- [x] Compiled successfully — split72, split42, and the monolithic `POLYKYBD_DOOM=yes` flavour (which PR CI does not build)
- [ ] Flashed and tested on hardware — **in progress**
- [x] If protocol changed: n/a, no protocol change

Measurements rather than impressions:

| | |
|---|---|
| RAM | **336 B**, measured on the monolith (`.heap` 2592 → 2256 B free) |
| `make test:polykybd_macro_record` | 47 tests, mutation-swept 7/7 |
| `make test:polykybd_macro_decode` | 23 tests |
| cppcheck | exit 0 (it caught a bound-order read in the legacy-caption compare) |
| lint job | `qmk lint --strict` both boards, `ci-validate-keyboard-targets`, `ci-validate-aliases`, tracked-ignored — all clean |
| status-OLED layout | widest ink 101 px of 128, tallest 14 px in a 16 px band, 0 clipped pixels, both panel heights |
| keycap legends | 0 off-panel pixels in both REC states, rendered through `oled_preview` and checked against the shipped `RESET`/`Eden` legend |
| stock marks | ink 4–19 px against 32 free rows, so none is halved and 0 pixels clip |

Paired with **PolyKybdHost#221** (the caption face ladder and the reshipped `symbol` bundle) and **polykybd-docs#73**.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MNyPVgtQG3iHzMjYntsWTz

---
_Generated by [Claude Code](https://claude.ai/code/session_01MNyPVgtQG3iHzMjYntsWTz)_

## Summary by CodeRabbit

- **New Features**
  - Added on-keyboard macro recording with slot selection, cancellation, saving, and capacity feedback.
  - Added OLED status screens showing recording state, selected slot, elapsed time, and saved size.
  - Added macro banking and expanded macro-key assignments on default layouts.
  - Added Mayan numeral icons and improved default labels for unclaimed macro slots.
  - Added a board-state generation indicator for detecting settings changes.

- **Documentation**
  - Added guidance for macro recording, display behavior, compatibility, and layout constraints.

- **Tests**
  - Added coverage for recording, report handling, storage updates, and edge cases.

### CODE REVIEW

File: keyboards/polykybd/base/macro_record.c [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+    uint16_t term = start;
+    while (term < end && rd(term, ctx) != 0u) {
+        term++;
+    }
+    // No terminator at all means the region is not a macro buffer (or is mid-write).
+    // Splicing into it would guess where the next macro starts.
+    if (term >= end) return POLY_MACRO_SPLICE_BAD_ID;
+
+    const uint16_t old_span = (uint16_t)(term + 1u - start);
+    const uint16_t new_span = (uint16_t)(len + 1u);
```
[the remark this answers is not in the document]
Comment (thpoll83) [reply keyboards/polykybd/base/macro_record.c]: Confirmed and fixed in `6cf5f25`.

Verified the mechanism before fixing: `POLY_MACRO_COMMIT_BODY` exits on `c->cursor > c->len`, and `cursor` is a `uint16_t` — so at `len == UINT16_MAX` it can never exceed it. The commit does not merely stall, it never terminates, and the buffer is left stamped `POLY_MACRO_INCOMPLETE`, i.e. permanently unplayable.

Your reachability caveat is right too — `poly_macro_rec_tick()` bounds its staging buffer to `POLY_MACRO_REC_BYTES` (192), so nothing gets there today. Taken anyway, because `len` is a `uint16_t` in this module's own contract and this file is the pure seam extracted precisely because the splice arithmetic has a bug history.

The span is now computed in `uint32_t` and refused above `UINT16_MAX`. `AnUnrepresentableSpanIsRefusedRatherThanWrapped` pins it, and was mutation-checked rather than assumed: deleting the one guard line fails that test and only that test (47 pass / 1 fail), restoring it gives 48/48.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: keyboards/polykybd/base/tests/macro_record_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,689 @@
...
+
+TEST(MacroSplice, TheBufferReadsNotIntactForTheWholeSplice) {
+    // A power cut mid-shift must leave something poly_macro_start() refuses. Otherwise
+    // the tail of a former macro is promoted into a macro of its own, and a keypress
+    // types a fragment of whatever it used to hold.
+    Region              r = make({tap(KC_A_), tap(KC_B_), tap(KC_C_)});
+    const uint16_t      end = (uint16_t)r.bytes.size();
+    poly_macro_commit_t c{};
+    ASSERT_EQ(poly_macro_commit_begin(&c, rd, wr, &r, 1,
+                                      (const uint8_t *)tap(KC_C_).data(), 3, end),
```
[the remark this answers is not in the document]
Comment (thpoll83) [reply keyboards/polykybd/base/tests/macro_record_tests.cpp]: Correct — fixed in `6cf5f25`.

`poly_macro_commit_begin()` only stores `c->body`; the `BODY` phase dereferences it many `poly_macro_commit_step()` calls later, by which point the `tap(KC_C_)` temporary is long dead. Genuine UB, and the neighbouring `TheChunkSizeDoesNotChangeTheResult` a few lines down already does the right thing (`const std::string body = ...`), so this was just inconsistent with the file's own idiom.

Now bound to a local `std::string` and passed with `body.size()` rather than a hardcoded 3.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: keyboards/polykybd/fonts/fonts.yaml [hunk 3]
```diff
@@ -31,6 +31,7 @@ sources:
   noto-sans-symbols:   fonts/Noto_Sans_Symbols/static/NotoSansSymbols-Regular.ttf
   noto-sans-symbols2:  fonts/Noto_Sans_Symbols_2/NotoSansSymbols2-Regular.ttf
   noto-sans-math:      fonts/Noto_Sans_Math/NotoSansMath-Regular.ttf
+  noto-sans-mayan:     fonts/Noto_Sans_Mayan_Numerals/NotoSansMayanNumerals-Regular.ttf
```
[the remark this answers is not in the document]
Comment (thpoll83) [reply keyboards/polykybd/fonts/fonts.yaml:34]: Declining this one — there is no directory here to rename.

`Noto_Sans_Mayan_Numerals/...` is a **download destination**, not a path in the repo. `dl-fonts.sh` fetches the TTF into a gitignored cache; `git ls-files 'keyboards/polykybd/**' | grep -c '\.ttf$\|\.otf$'` returns **0**, so this PR adds no directory and no file that the lowercase guideline governs.

The string also mirrors google/fonts' own `ofl/` layout, and it matches the convention its neighbours in this very block already use — `Noto_Emoji`, `Noto_CEmoji`, `Noto_Sans_Symbols`, `Noto_Sans_Symbols_2` and, immediately above the new line, `Noto_Sans_Math`. Renaming one of six would make the file less consistent, not more.

And it is not local: `keyboards/polykybd/fonts/noto-fonts.yaml` is kept **byte-identical** with `PolyKybdHost/polyhost/res/fonts/noto-fonts.yaml` (the host's "Download Noto…" button reads the same catalog). A rename would be a two-repo change, verified with `cmp`, for no gain.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: CLAUDE.md [hunk 4]
```diff
@@ -1642,6 +1642,82 @@ keycode; `process_record_user()` calls it last, before `display_wakeup()`.
   (810) only changes the identity's **extent**: indices `< NUM_OVERLAY_SLOTS` are
   identity, the rest are a 0 fill that can never be an upload destination.
 
+### Telling the host something changed ON THE BOARD
```
[the remark this answers is not in the document]
Comment (thpoll83) [reply CLAUDE.md:1645]: Declining — the contract being applied here is upstream QMK's, and it does not govern this repository.

This is a fork (`thpoll83/qmk_firmware`), and the PR targets its own default branch `PolyKybd`, not `qmk/qmk_firmware`. There is no QMK Collaborator in this review path to defer to.

`CLAUDE.md` is also not an upstream file — `curl` against `raw.githubusercontent.com/qmk/qmk_firmware/master/CLAUDE.md` returns **404**. It is this fork's own guidance file, written and maintained here, and the PolyKybd convention is that a change to firmware behaviour updates it in the same PR: the notes are how the next session learns why the code is shaped the way it is. Deferring it would leave the file describing a keyboard that no longer matches.

Worth flagging for the ruleset rather than for this PR: the `keyboards/**/*` routing rule is a reasonable default for upstream, but in this fork almost every change is PolyKybd's and `CLAUDE.md` travels with it, so the rule will fire on most PRs here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: keyboards/polykybd/poly_keymap.c [hunk 5]
```diff
@@ -4093,6 +4230,51 @@ bool process_record_user(uint16_t keycode, keyrecord_t* record) {
...
+    if (poly_macro_rec_state() == POLY_REC_PICKING) {
+        if (!record->event.pressed) {
+            const uint16_t ul = macro_picker_keycode_at(record->event.key.row,
+                                                        record->event.key.col);
+            if (ul >= QK_MACRO && ul <= QK_MACRO_MAX) {
+                const uint8_t id = poly_macro_banked_id((uint8_t)(ul - QK_MACRO),
+                                                        (get_mods() & MOD_MASK_SHIFT) != 0);
+                if (id != POLY_MACRO_NONE) poly_macro_rec_pick(id);
+            } else if (ul == KC_MACRO_REC) {
+                poly_macro_rec_toggle();   // the picker's own way out
```
[the remark this answers is not in the document]
Comment (thpoll83) [reply keyboards/polykybd/poly_keymap.c:4286]: Confirmed and fixed in `8b630c3`. This was a real defect I introduced in `2a9cd9a5`, and it would have shipped — the `.bin` already handed over for hardware testing has it, so thank you, this saved a wasted round.

Traced before fixing: the opening press is handled by the `KC_MACRO_REC` block one level down **while the state is still `POLY_REC_IDLE`**, so the picker branch never sees it. By the time the matching release arrives the state is `POLY_REC_PICKING`, the branch matches first, `macro_picker_keycode_at()` resolves the position to `KC_MACRO_REC`, and `poly_macro_rec_toggle()` takes `PICKING -> enter_idle()`. One tap opens and closes it, which on hardware reads as "REC does nothing".

Fixed one level more generally than "ignore the opening REC release": the picker now answers **only a release whose press it saw itself**, tracked as a single `(row, col)` and armed to `POLY_PICK_NO_PRESS` at the one place the picker is opened. So the opening release is ignored because its press happened while the picker was closed, and the same rule covers any other key whose press predates the mode — rather than special-casing one keycode. A later REC tap still cancels on its release, which keeps the reason that edge was chosen (`matrix_scan_kb` inverts a keycap on press and un-inverts on release independently of `process_record`, so acting on the press tears the picker down while the cap is still inverted).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: keyboards/polykybd/poly_keymap.c [hunk 6]
```diff
@@ -4093,6 +4230,51 @@ bool process_record_user(uint16_t keycode, keyrecord_t* record) {
...
+            }
+        }
+        return false;
```
[the remark this answers is not in the document]
Comment (thpoll83) [reply keyboards/polykybd/poly_keymap.c:4289]: Confirmed and fixed in `8b630c3`.

The chain is exactly as described: `process_record_user()` returning false stops `process_record()` before `process_action()`, so a swallowed Shift never registers, `get_mods()` never gains `MOD_MASK_SHIFT`, and `poly_macro_banked_id(slot, shift)` can only ever return M0..M11. The render path is hit by the same cause from the other side — it reads the **synced** `local_layer->mods`, which is fed from the master's live mods, so the picker would not even have drawn M12..M15 to aim at.

`IS_MODIFIER_KEYCODE(keycode)` now returns true from the picker branch so the event reaches QMK; everything else is still swallowed. Safe on the host for the same reason the Intl picker's Ctrl latch is: a bare modifier tap does nothing in the apps this has to survive.

On the `clear_keyboard()` half — kept, and the consequence documented rather than worked around. It has to drop a held modifier before the mode changes (the picker swallows the release of a key that was already down, so the host would auto-repeat it), so a Shift held *through* the REC tap is gone and has to be pressed again once the picker is up. That is the gesture as intended — tap REC, then hold Shift and tap a macro key — and re-pressing works because the modifier now passes through.

---
_Generated by [Claude Code](https://claude.ai/code)_
